tests/provider: Initial goreleaser configuration and CI workflow

bflad opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #8988
Reference: https://www.terraform.io/docs/registry/providers/publishing.html

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
NONE
```

Going forward, the Terraform Provider release process for the Terraform Registry will recommend `goreleaser` for all providers, including the official ones released via a separate process today. Given this decision is now more final, we can begin the process of using this tool for CI. In the future, we can also support nightly builds and later releases via the new GitHub release process. For now though, we leverage it just to verify codebase cross-compilation for all platforms that will be supported after the upcoming 3.0.0 release.

Output from acceptance testing: N/A (CI testing)

radeksimko commented on .github/workflows/terraform_provider.yml:
```diff
@@ -225,6 +226,29 @@ jobs:
+      - name: goreleaser check
+        uses: goreleaser/goreleaser-action@v2
+        with:
+          args: check
```
FYI: GoReleaser will return non-zero exit code here when it detects deprecated options in the config, which is probably good to catch either way, but you might still want to `continue-on-error` in such case as the config may otherwise be valid and the next step can still check if it builds on all platforms.

radeksimko commented on .goreleaser.yml:
```diff
@@ -0,0 +1,42 @@
+builds:
+  - binary: '{{ .ProjectName }}_{{ .Version }}'
+    env:
+      - CGO_ENABLED=0
```
Not sure if `go mod download` does anything that this may affect, but you might want to move it to the top-level `env` section, so that any other `go` commands outside of `builds` can pick it up too.

radeksimko commented on .goreleaser.yml:
```diff
@@ -0,0 +1,42 @@
+archives:
+  - name_template: '{{ .ProjectName }}_{{ .Version }}_{{ .Os }}_{{ .Arch }}'
+    format: zip
```
GoReleaser packages `README` and `LICENSE` files by default here, which you may not want - at least I think we haven't packaged anything other than binaries so far for any providers.

```yaml
files: # Avoid packaging any extra (default) files
      - none*
```

bflad commented:
Thanks, @radeksimko! Adjusted per your recommendations. Also going to see if re-enabling build concurrency is a good or bad idea, since the timeout is now fixed to match GitHub Actions'.

paultyng commented on .goreleaser.yml:
```diff
@@ -0,0 +1,44 @@
+      - goarch: '386'
+        goos: darwin
+    ldflags:
+      - -s -w -X github.com/terraform-providers/terraform-provider-aws/aws/version.ProviderVersion={{.Version}}
```
You are probably going to want `-trimpath` for `flags` here as well as overriding the `mod_timestamp`, see these very helpful instructions an awesome contributor put together for Goreleaser: https://goreleaser.com/customization/build/#reproducible-builds

bflad replied:
Is https://github.com/terraform-providers/terraform-provider-aws/pull/14033/commits/aa78000afa74380bb785fae709c8a4559a9d2f49 what you had in mind here? 😄

paultyng replied:
Yep, I can't really speak to the ldflags -X change, but the other parts look good.
